MapboxNavigation/NavigationMapView.swift

```diff
@@ -410,55 +400,36 @@ open class NavigationMapView: MGLMapView, UIGestureRecognizerDelegate {
         
     }
     
-    //TODO: Change to point-based distance calculation
-    private func waypoints(on routes: [Route], closeTo point: CGPoint) -> [Waypoint]? {
-        let tapCoordinate = convert(point, toCoordinateFrom: self)
-        let multipointRoutes = routes.filter { $0.routeOptions.waypoints.count >= 3}
-        guard multipointRoutes.count > 0 else { return nil }
-        let waypoints = multipointRoutes.flatMap({$0.routeOptions.waypoints})
+    @objc func updateCourseView(_ sender: UIGestureRecognizer) {
+        preferredFramesPerSecond = FrameIntervalOptions.defaultFramesPerSecond
         
-        //lets sort the array in order of closest to tap
-        let closest = waypoints.sorted { (left, right) -> Bool in
-            let leftDistance = left.coordinate.distance(to: tapCoordinate)
-            let rightDistance = right.coordinate.distance(to: tapCoordinate)
-            return leftDistance < rightDistance
+        if sender.state == .ended {
+            altitude = self.camera.altitude
+            enableFrameByFrameCourseViewTracking(for: 2)
         }
         
-        //lets filter to see which ones are under threshold
-        let candidates = closest.filter({
-            let coordinatePoint = self.convert($0.coordinate, toPointTo: self)
-            return coordinatePoint.distance(to: point) < tapGestureDistanceThreshold
-        })
-        
-        return candidates
-    }
-    
-    private func routes(closeTo point: CGPoint) -> [Route]? {
-        let tapCoordinate = convert(point, toCoordinateFrom: self)
-        
-        //do we have routes? If so, filter routes with at least 2 coordinates.
-        guard let routes = routes?.filter({ $0.coordinates?.count ?? 0 > 1 }) else { return nil }
+        // Capture altitude for double tap and two finger tap after animation finishes
+        if sender is UITapGestureRecognizer, sender.state == .ended {
```

Comment:
This is way we should use `touchesEnded(_: ,with)`. With that, we can use  Set<UITouch>.count() to determine the number of fingers the user is using and we are guaranteed touches did end. (PS: I don't see where we're setting `numberOfTapsRequired` on our gesture recognizer to `2` either. So this is certainly a concern.)

Reply:
**Follow up**: So it makes sense to use `UITapGestureRecognizer`, however, I'm still concerned that we're not setting `numberOfTapsRequired` to 2 on our gesture recognizer but comment reads that `Capture altitude for double tap and two finger tap after animation finishes`... thoughts?

Reply:
Not necessarily disagreeing with you, but it's worth noting that this code was pulled wholesale from [here.](https://github.com/mapbox/mapbox-navigation-ios/pull/1063/files/3d54668965e7ba331c280d31a10a9a62b3c79b33#diff-a43bde32226ca771fb8f732e90f4fc6aL213)